plugins/module_utils/structure_merging.py

```diff
@@ -0,0 +1,250 @@
+            raise TypeError(msg.format(type(base), type(changes)))
+        self._base = base
+        self._changes = changes
+        self._present = present
+        self._merge_seq_by_index = merge_seq_by_index
+        self._keep_empty = keep_empty
+        self._remove_none = remove_null
+
+    def get(self):
+        # type: () -> dict | list
+        """
+        Get the new structure depending on initialisation parameters.
+        """
+        return self._get_method(self._base, self._changes)
+
+    def _get_new_dict_with_changes_present(self, base, changes):
+        # type: (Mapping, Mapping) -> dict
+        new = {}
+        keys_base = base.keys()
+        keys_changes = changes.keys()
+        keys_all = list(keys_base) + [key for key in keys_changes if key not in keys_base]
+        for k in keys_all:
+            # Ignoring value of changes[k] when it None permit to avoid to
+            # update this item when it explicitly skipped with
+            # `merge_seq_by_index=True`.
+            if k not in keys_changes or changes[k] is None:
+                new[k] = base[k]
+            elif self._both_are_assimilable_to_dict(base.get(k), changes[k]):
+                new[k] = self._get_new_dict_with_changes_present(base[k], changes[k])
+            elif self._both_are_assimilable_to_list(base.get(k), changes[k]):
+                new[k] = self._get_new_list_with_changes_present(base[k], changes[k])
+            else:
+                new[k] = changes[k]
+        return new
+
+    def _get_new_dict_with_changes_absent(self, base, changes):
+        # type: (Mapping, Mapping) -> dict
+        new = {}
+        for k, base_elem in base.items():
+            changes_elem = changes.get(k)
+            if self._both_are_assimilable_to_dict(base_elem, changes_elem):
+                res = self._get_new_dict_with_changes_absent(base_elem, changes_elem)
+                if len(res) > 0 or self._keep_empty:
+                    new[k] = res
+            elif self._both_are_assimilable_to_list(base_elem, changes_elem):
+                res = self._get_new_list_with_changes_absent(base_elem, changes_elem)
+                if len(res) > 0 or self._keep_empty:
+                    new[k] = res
+            elif not base_elem == changes_elem and not (
+                    self._remove_none and changes_elem is None and k in changes.keys()):
+                new[k] = base_elem
+        return new
+
+    def _get_new_list_with_changes_present(self, base, changes):
+        # type: (Sequence, Sequence) -> list
+        if self._merge_seq_by_index:
+            base_dict = self._convert_generic_sequence_to_dict(base)
+            changes_dict = self._convert_generic_sequence_to_dict(changes)
+            res_dict = self._get_new_dict_with_changes_present(base_dict, changes_dict)
+            new = self._convert_generic_mapping_to_list(res_dict)
```

Comment:
Converting back and forth has a couple of problems there:
* increased memory usage (to hold the same data), which might become a problem if someone decides to use this for larger datasets
* lists, by definition, maintain order, and the conversion method back to list does not do that

Reply:
It could be solved by something (untested) like:
```suggestion
            for base_elem, chg_elem in zip(base, changes):
                res_elem = self._choose_merge_method(base_elem, chg_elem, present=True)(base_elem, chg_elem)
                res.append(res_elem)
```